src/crypto.rs

```diff
@@ -0,0 +1,41 @@
+use sha2::{Digest as _, Sha256};
+
+use crate::error::Result;
+use std::io::{BufRead as _, BufReader, Read};
+
+#[expect(dead_code, reason = "temporary for next pull request")]
+/// Size of the output that the crypto function spit out
+pub static HASH_SIZE_IN_BYTES: usize = 32;
+
+#[expect(dead_code, reason = "temporary for next pull request")]
+/// Function to hash data from a ``BufReader``
+///
+/// # Errors
+/// Will error out if failed to fill buffer for some reason
+pub fn hash_buf_reader<R: Read>(mut reader: BufReader<R>) -> Result<String> {
+    let mut hasher = Sha256::new();
+
+    loop {
+        let buffer_len = {
+            let buffer = reader.fill_buf()?;
+            hasher.update(buffer);
+            buffer.len()
+        };
+
+        if buffer_len == 0 {
+            break;
+        }
+        reader.consume(buffer_len);
+    }
+
+    Ok(format!("{:x}", hasher.finalize()))
+}
```

Comment:
Is the `BuffReader` necessary? Why not just use `read_exact` directly?

```suggestion
pub fn hash_stream(stream: &mut impl Read) -> Result<String> {
    const SIZE: usize = 32;
    let mut hash = Sha256::new();
    let mut buffer: [u8; SIZE];
    while let Ok(_) = {
        buffer = [0; SIZE];
        stream.read_exact(&mut buffer)
    } {
        hash.update(buffer);
    }
    hash.update(buffer
        .into_iter()
        .filter(|x| *x > 0)
        .collect::<Vec<_>>());
    Ok(format!("{:x}", hash.finalize()))
}
```

Reply:
Wasn't aware of this function, I can look into it

Reply:
Okay a quick look over it
```while let Ok(_) = {
        buffer = [0; SIZE];
        stream.read_exact(&mut buffer)
    } {
        hash.update(buffer);
    }
```

This doesn't exactly hash by chunks as I wanted it to. Basically the flow I was aiming for is read in 32 bytes -> update hash -> trash it and read the next 32 bytes. This allows it to read really big files without running into memory issues.

Thought the while let is new to me, and might be able to use that if it makes the code simpler

Reply:
Wait, isn't what I proposed doing the same thing? I allocated `buffer` and then the `read_exact` chunk overrides the previous value while updating the hash in-between i.e. discarding the prior value. Is it growing in memory in your test? That would be unexpected.

Reply:
I see, I got confused with line 26-28. Didn't realized that syntax was even a thing. I reading it as the hash.update part doesn't run every loop

Reply:
Okay I I haven't got a chance to look at one issue into futher, but the hashing comes out differently possibbly due to this behavior
```
If this function encounters an “end of file” before completely filling the buffer, it returns an error of the kind [ErrorKind::UnexpectedEof](https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.UnexpectedEof). The contents of buf are unspecified in this case.
```

https://doc.rust-lang.org/std/io/trait.Read.html#method.read_exact

I haven't gotton a chance to look into too deep, but that is so far my guess why it is resulting in a different hash.

Reply:
I created a version in the [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=0df55b82bc80afa04c79d4b03e9d557a) that compares them. All hashed to the name value.

Also, I'm going to conveniently do some hand-waving around Rust Playground's security. :upside_down_face:

Reply:
Hmm okay then it might be something on my side. Will check when I get back home